_posts/2017-06-15-REST-api-NodeJS-koa.md

```diff
@@ -0,0 +1,230 @@
+
+{% highlight coffeescript %}  
+    import { Container } from "TypeScript-ioc";
+    
+    class Foo {
+        doSomething(): string {
+        	return "Hello World";
+        }	
+    }
+    
+    class Bar {
+        constructor(@Inject private foo: Foo) {
+        }	
+    
+    	doAnotherThing(): string {
+    		this.foo.doSomething();
+        }
+    }
+    
+    const bar:Bar = Container.get(Bar);
+    bar.doAnotherThing();
+{% endhighlight %}
+
+### typeorm
+
+At first I just saved the movies and directors in the services as an in-memory array for testing purposes, but in a real application you will want persistence of some sort, so I needed a database. I decided on a regular old mysql database and an ORM library to do the mapping between the database records and my TypeScript model classes. For ORM I used [typeorm](https://github.com/typeorm/typeorm). It’s pretty easy to use. It also uses the annotations like TypeScript-ioc, which makes code very readable. The experience with this library was more or less pain free, so I really recommend it. To check a real example from my repository, check the [Movie](https://github.com/ivarvh/movielistr-backend-ts-ioc/blob/master/src/models/Movie.ts) model.
+
+# Testing
+
+### Unit testing
+
+For unit testing I used the classic combination of [Mocha](https://mochajs.org/), [Sinon](http://sinonjs.org/) and [Chai](http://chaijs.com/). Since I was using dependency injection, I also needed a good way of mocking my dependencies, for this I found ts-mockito. Ts-mockito is more or less a clone of the mockito library in Java. It allows you to create mocks of classes, make functions return certain values and verify that calls have been made. This made it super easy to write tests. For examples check the tests folder in my repository. To execute the TypeScript tests, I used [ts-node](https://github.com/TypeStrong/ts-node). Ts-node compiles the TypeScript and keeps the compiled JavaScript in memory while it executes it. This way you don’t have to create an additional folder to compile the tests to and execute them. You can then easily create an npm script like this:
```

Comment:
You could make ts-mockito link to the package (https://www.npmjs.com/package/ts-mockito).